File: code/python/interview_arena_article-05.py
"""Thinking framework when actually using binary search"""

from __future__ import annotations
from typing import Any


def solve(*args: Any, **kwargs: Any) -> Any:
    """Run this lesson exercise with positional inputs."""
    return _Solution().shipWithinDays(*args, **kwargs)


class _Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        left = max(weights)
        right = sum(weights)
        while left < right:
            mid = (left + right)
            if self.f(weights, mid) <= days:
                right = mid
            else:
                left = mid + 1
        return left
    def f(self, weights, x):
        days = 0
        i = 0
        while i < len(weights):
            cap = x
            while i < len(weights):
                if cap < weights[i]:
                    break
                else:
                    cap -= weights[i]
                    i += 1
            days += 1
        return days

File: code/java/interview_arena_article-05.java
public class Interview_arena_article_05 {
    /**
     * Thinking framework when actually using binary search
     */
    public Object solve(Object... args) {
        if (args == null || args.length != 2) {
            throw new IllegalArgumentException("Expected: (int[] weights, int days)");
        }

        Solution solution = new Solution();
        return solution.shipWithinDays((int[]) args[0], (int) args[1]);
    }

    private static class Solution {

        public int shipWithinDays(int[] weights, int days) {
            int left = 0;
            int right = 1;
            for (int w : weights) {
                left = Math.max(left, w);
                right += w;
            }
            while (left < right) {
                int mid = left + (right - left) / 2;
                if (f(weights, mid) <= days) {
                    right = mid;
                } else {
                    left = mid + 1;
                }
            }
            return left;
        }
        int f(int[] weights, int x) {
            int days = 0;
            for (int i = 0; i < weights.length; ) {
                int cap = x;
                while (i < weights.length) {
                    if (cap < weights[i]) break;
                    else cap -= weights[i];
                    i++;
                }
                days++;
            }
            return days;
        }
    }
}
